pkg/controller/controller_test.go

```diff
@@ -0,0 +1,259 @@
+package controller
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+
+	"k8s.io/client-go/kubernetes"
+
+	"k8s.io/apimachinery/pkg/watch"
+
+	"k8s.io/klog"
+
+	"k8s.io/client-go/rest"
+
+	"github.com/golang/mock/gomock"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/submariner-io/admiral/pkg/federate/mocks"
+	lighthousev1 "github.com/submariner-io/lighthouse/pkg/apis/lighthouse.submariner.io/v1"
+	core_v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	fakeCoreClientSet "k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/testing"
+)
+
+type watchReactor struct {
+	multiclusterServiceWatchStarted chan bool
+}
+
+type eqMultiClusterService struct {
+	expected *lighthousev1.MultiClusterService
+}
+
+var _ = Describe("Lighthouse Controller", func() {
+	klog.InitFlags(nil)
+
+	When("start is called", testStart)
+	Describe("Cluster lifecycle notifications", testClusterLifecycleNotifications)
+	Describe("Lighthouse resource distribution", testResourceDistribution)
+})
+
+func testStart() {
+	var (
+		mockCtrl      *gomock.Controller
+		mockFederator *mocks.MockFederator
+	)
+
+	BeforeEach(func() {
+		mockCtrl = gomock.NewController(GinkgoT())
+		mockFederator = mocks.NewMockFederator(mockCtrl)
+	})
+
+	AfterEach(func() {
+		mockCtrl.Finish()
+		mockCtrl = nil
+		mockFederator = nil
+	})
+
+	When("WatchClusters succeeds", func() {
+		It("should return no error", func() {
+			mockFederator.EXPECT().WatchClusters(gomock.Any())
+			Expect(New(mockFederator).Start()).To(Succeed())
+		})
+	})
+
+	When("WatchClusters fails", func() {
+		It("should return an error", func() {
+			mockFederator.EXPECT().WatchClusters(gomock.Any()).Return(errors.New("mock error"))
+			Expect(New(mockFederator).Start()).ToNot(Succeed())
+		})
+	})
+}
+
+func testClusterLifecycleNotifications() {
+	var (
+		controller   *LightHouseController
+		watchReactor *watchReactor
+	)
+
+	BeforeEach(func() {
+		controller = New(mocks.NewMockFederator(gomock.NewController(GinkgoT())))
+		watchReactor = newWatchReactor(controller)
+	})
+
+	AfterEach(func() {
```

Comment:
perhaps niling again like you did in the previous test?

Reply:
Done.